README.md
# Dalla Stats ![](https://img.shields.io/github/release/egeldenhuys/dalla-stats.svg)

Scrape traffic statistics from TP-LINK AC750 Archer C20 Router

## Usage

```
usage: dalla-stats.py [-h] [-u USERNAME] [-p PASSWORD] [-i INTERVAL]
                      [-d ROOT_DIRECTORY] [-l] [-v]

optional arguments:
  -h, --help            show this help message and exit
  -u USERNAME, --username USERNAME
                        the router admin username
  -p PASSWORD, --password PASSWORD
                        the router admin password
  -i INTERVAL, --interval INTERVAL
                        the interval in seconds to update the statistics.
  -d ROOT_DIRECTORY, --root-directory ROOT_DIRECTORY
                        directory to save logs
  -l, --disable-logging
                        Disable logging of statistics
  -v, --version         show program's version number and exit
```


CHANGELOG.md
# Changelog

This project adheres to [Semantic Versioning](http://semver.org/).

## Unreleased
- Create simple logger from dalla-stats.py

## [0.2.3] - 2016-11-01
### Fixed
- Remove testing code breaking normal functionality

## [0.2.2] - 2016-10-31 [YANKED]
### Fixed
- Fix crash when new day starts #51

## [0.2.1] -2016-10-31
### Added
- Force logging when new day starts #50

## [0.2.0] - 2016-10-31
### Added
- Ability to output Delta per second to console/log file (Disabled) #31
- Summary is now ranked from highest to lowest total usage #46
- Percentage of data used is now displayed next to each user #35

### Changed
- Poll for new records every 1 second (better accuracy of records), and write logs at --interval seconds #38
- If the log and poll interval are not the same, then the Delta in the log files
is no longer accurate and should not be sued for calculations
- Human readable timestamp when logging the get records function
- TOTAL is now a user #23
- User 'Unkown' has been renamed to 'UNKNOWN'. File needs to be renamed as well!

### Fixed
- Daily stats now start at 0 if it is the first run

## [0.1.2+dev.1] - 2016-10-30
### Added
- Summary is now ranked from highest to lowest total usage #46
- Percentage of data used is now displayed next to each user #35

### Changed
- TOTAL is now a user #23
- User 'Unkown' has been renamed to 'UNKNOWN'. File needs to be renamed as well!

### Fixed
- Daily usage will now be correct on first run.

## [0.1.2] - 2016-10-29
## Fixed
- Day will now increment when it is a new day

## [0.1.1] - 2016-10-29
### Fixed
- Reset the daily statistics when a new day starts

## [0.1.0] - 2016-10-29
### Added
- Reset usage when a new month begins and start a new log folder #16
- Reset byte buffer when device not found on router #30
- Produces HTML Total, and Today pages #36, #24, #17

### Changed
- File paths are now hard-coded in main, instead of in the separate functions #15
- Detailed log when negative delta #28
- Change directory structure to allow monthly captures #33
- Logging is now enabled by default
- Default interval has been set to 60 seconds
- --interval, --root-directory and --disable-logging are now optional
- Better WARN log style

### Fixed
- CRTL+C losing data while waiting for router
- Not logging out when failling to get records from router #44

## [0.0.5] - 2016-10-27
### Fixed
- Duplicate log entries when device not found on router #41

## [0.0.4] - 2016-10-25
### Added
- Log timestamp to console when fetching records
- Now classify data has On-Peak or Off-Peak depending on timestamp of device record #3
- Create overview.csv that contains a summary for TOTAL and each USER #5
- Catches KeyboardInterrupt and does a final capture and saves to disk #12

## [0.0.3] - 2016-10-24
### Added
- Port to Python 3

### Fixed
- Handles router errors #10
- Keeps data on restart #11

## [0.0.2] - 2016-10-24
### Added
- Store a summary of the device records on disk #2, #4
- Monitor only mode (Produces no .csv for graphing) #8
- Will now place all log files in the given working directory

### Removed
- Remove graphs from README.md #13
- Remove numpy dependency (was not working on Termux)

## 0.0.1 - 2016-10-23
- Initial release

[Unreleased]: https://github.com/egeldenhuys/dalla-stats/compare/v0.2.3...HEAD
[0.2.3]: https://github.com/egeldenhuys/dalla-stats/compare/v0.2.2...v0.2.3
[0.2.2]: https://github.com/egeldenhuys/dalla-stats/compare/v0.2.1...v0.2.2
[0.2.1]: https://github.com/egeldenhuys/dalla-stats/compare/v0.2.0...v0.2.1
[0.2.0]: https://github.com/egeldenhuys/dalla-stats/compare/v0.1.2...v0.2.0
[0.1.2+dev.1]: https://github.com/egeldenhuys/dalla-stats/compare/v0.1.1...v0.1.2+dev.1
[0.1.2]: https://github.com/egeldenhuys/dalla-stats/compare/v0.1.1...v0.1.2
[0.1.1]: https://github.com/egeldenhuys/dalla-stats/compare/v0.1.0...v0.1.1
[0.1.0]: https://github.com/egeldenhuys/dalla-stats/compare/v0.0.5...v0.1.0
[0.0.5]: https://github.com/egeldenhuys/dalla-stats/compare/v0.0.4...v0.0.5
[0.0.4]: https://github.com/egeldenhuys/dalla-stats/compare/v0.0.3...v0.0.4
[0.0.3]: https://github.com/egeldenhuys/dalla-stats/compare/v0.0.2...v0.0.3
[0.0.2]: https://github.com/egeldenhuys/dalla-stats/compare/v0.0.1...v0.0.2


dalla-logger/dalla-logger.py
import argparse
import requests
import base64
import time
import os
from os import listdir
from os.path import isfile, join
import csv
import datetime
import logging
import sys
import calendar


version = 'v0.0.1'

def main():

	print('[{0}] [INFO] Starting Dalla-Data {1}'.format(getTime(), version))

	parser = argparse.ArgumentParser()

	parser.add_argument("-u", "--username", default='', help="the router admin username")
	parser.add_argument("-p", "--password", default='', help="the router admin password")
	parser.add_argument("-i", "--interval", type=int, default=60, help="the interval in seconds to update the log files")
	parser.add_argument("-d", "--device-directory", default='', help="directory to save device logs")
	#parser.add_argument("-l", "--service-log-file", default='', help="directory to save device logs")
	parser.add_argument('-v', '--version', action='version', version='%(prog)s ' + version)

	args = parser.parse_args()

	deviceDir = args.device_directory

	if deviceDir == '':
		print('Please provide a directory for the data logs with -d')
		exit()

	interval = args.interval

	if (deviceDir == ''):
		deviceDir = 'devices'

	if (args.username == '' or args.password == ''):
		print('[ERROR] Please supply username and password')
		exit()

	if (interval == 0):
		print('[ERROR] Interval needs to be > 0')
		exit()

	sessionHeaders = initSession(args.username, args.password)

	abort = False
	while (True):
		try:
			timeKey = int(time.time()) # UTC TIME!

			devices = {}

			print('[{0}] [INFO] Fetching device stats...'.format(getTime()))
			devices = getDeviceRecords(sessionHeaders, timeKey)

			if (len(devices) != 0):
				print('[{0}] [INFO] Logging device stats to log'.format(getTime()))
				logDeviceStats(devices, deviceDir)
			else:
				print('[{0}] [ERROR] Failed to get device records from router.'.format(getTime()))
				# Getting records fail, try to logout
				# REVIEW: Does this work when timout occurs
				logout(sessionHeaders)

			if (abort == False):
				time.sleep(interval)
			else:
				break

		except (KeyboardInterrupt, SystemExit) as e:
			print(e)
			print('\n[{0}] [INFO] Exiting. Please wait...'.format(getTime()))
			time.sleep(1)
			abort = True

def getTime():
	timeKey = int(time.time()) # UTC TIME!
	localTimeFormated = time.strftime('%Y-%M-%d %H:%M:%S', time.localtime(timeKey))
	return localTimeFormated

def validateDevices(statsDictArray, timeKey):
	"""Strip invalid entries and parse valid entries
	"""

	newDictArray = []

	for statsDict in statsDictArray:
		# Skip invalid intries
		if (len(statsDict) == 12):
			# Strip extra data,
			# Convert Dec IP to readable IP
			tmpDict = { 'MAC Address': statsDict['macAddress'],
						'IP Address': decStrToIpStr(statsDict['ipAddress']),
						'Total Bytes': int(statsDict['totalBytes']),
						'Time': timeKey}

			newDictArray.append(tmpDict)

	return newDictArray


def getDeviceRecords(session, timeKey):
	""" Poll the router for the current device statistics

	These records need to be compared to a previous set to calculate the
	Delta
	"""

	# Configure page specific headers
	url = 'http://192.168.1.1/cgi?1&5'
	session.headers.update({'Referer': 'http://192.168.1.1/mainFrame.htm'})
	data ='[STAT_CFG#0,0,0,0,0,0#0,0,0,0,0,0]0,0\r\n[STAT_ENTRY#0,0,0,0,0,0#0,0,0,0,0,0]1,0\r\n'

	try:
		r = session.post(url=url, data=data, timeout=1)
	except requests.ConnectionError:
		print('[ERROR] Network unreachable!')
		return {}
	except requests.ReadTimeout:
		print('[ERROR] Connection timeout!')
		return {}
	except KeyboardInterrupt:
		#print('[ERROR] KeyboardInterrupt during getDeviceRecords()')
		raise
	except:
		print('[ERROR] Unexpected error: ', sys.exc_info()[0])
		return {}

	rawStats = r.text

	error = rawStats.split('\n')

	if (error[-1] != '[error]0'):
		print('[{0}] [ERROR] Failed to get device records from router!'.format(getTime()))
		if (r.text == '<html><head><title>500 Internal Server Error</title></head><body><center><h1>500 Internal Server Error</h1></center></body></html>'):
			print('[{0}] Another admin has logged in!'.format(getTime()))
		else:
			print('\t' + r.text)

		return []

	dictArray = []
	tmpDict = {}

	arr = rawStats.split("\n")

	for i in range(0, len(arr)):
		"""
		Loop through every line
		If the line is a title
			begin split and read into dict

		if we encounter another title
			insert old dict into array and start new dict and increase index

		"""

		arr[i] = arr[i].strip()

		# start of a new header
		if (arr[i][0] == '['):

			dictArray.append(tmpDict)
			tmpDict = {}
			next # skip the header

		tmp = arr[i].split('=')

		# Add key=value
		if (len(tmp) == 2):
			tmpDict[tmp[0]] = tmp[1]

	# Manipulate dict array to get what we need
	init = validateDevices(dictArray, timeKey)

	logout(session)

	return init

def logDeviceStats(statsDictArray, deviceDir):
	"""Save device dict array to log files
	"""
	"""
	filename: M-A-C_I.P
	Time, Total Bytes
	"""

	if (not os.path.exists(deviceDir)):
		os.makedirs(deviceDir)

	for statsDict in statsDictArray:

		# Generate file name
		mac = statsDict['MAC Address'].replace(':', '-')
		ip = statsDict['IP Address']
		fileName = str(deviceDir + '/' + mac + '_' + ip + '.csv')

		# csv fields
		timeKey = statsDict['Time']
		totalBytes = statsDict['Total Bytes']

		# new device, set up csv for it
		header = False
		if (os.path.isfile(fileName) == False):
			header = True

		output = open(fileName, 'a')

		if (header):
			output.write('Time, Total Bytes\n')

		output.write('{0}, {1}\n'.format(timeKey, totalBytes))
		output.close()


def getIpFromFileName(name):
	result = name.split('_')
	result = str(result[1])[:-4]

	return result

def getMacFromFileName(name):
	result = name.split('_')
	result = result[0].replace('-', ':')

	return result

def initSession(username, password):
	session = requests.session()

	raw = username + ':' + password
	encoded = base64.b64encode(raw.encode('utf-8'))

	auth = 'Basic ' + encoded.decode('utf-8')
	cookie = 'Authorization=' + auth

	session.headers = {
		'Host': '192.168.1.1',
		'User-Agent': 'Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:46.0) Gecko/20100101 Firefox/46.0',
		'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8',
		'Accept-Language': 'en-US,en;q=0.5',
		'Accept-Encoding': 'gzip, deflate',
		'Content-Type': 'text/plain',
		'Cookie': cookie,
		'Referer': 'http://192.168.1.1/',
		'Connection': 'keep-alive'
	}

	return session

def decStrToIpStr(dec):
	binStr = bin(int(dec))

	binStr = binStr[2:]
	finalStr = ''

	"""
	0: 0, 8 (8*0), ()
	1: 8, 16 (8*1)
	2: 16, 24 (8*2)
	3: 24, 32 (8*3)
	"""

	for i in range(0, 4):
		tmp = binStr[8 * i : 8 * (i + 1)]
		tmp = int(tmp, 2)
		finalStr += str(tmp) + '.'

	return finalStr[:-1]


def logout(session):
	# Configure page specific headers
	url = 'http://192.168.1.1/cgi?8'
	session.headers.update({'Referer': 'http://192.168.1.1/MenuRpm.htm'})
	data ='[/cgi/logout#0,0,0,0,0,0#0,0,0,0,0,0]0,0\r\n'

	try:
		r = session.post(url=url, data=data)
	except KeyboardInterrupt:
		#print('[ERROR] KeyboardInterrupt during getDeviceRecords()')
		raise
	except:
		print('[{0}] [ERROR] Unexpected error: {1}'.format(getTime(), sys.exc_info()[0]))
		return


	if (r.text != '[cgi]0\n[error]0'):
		print('[{0}] [ERROR] Logout failed:'.format(getTime()))

		if (r.text == '<html><head><title>500 Internal Server Error</title></head><body><center><h1>500 Internal Server Error</h1></center></body></html>'):
			print('[{0}] Another admin has logged in!'.format(getTime()))
		else:
			print('\t' + r.text)


main()
